topics/tutorials/postfix_completion/postfix_templates.md

```diff
@@ -0,0 +1,77 @@
+
+> See the [](advanced_postfix_templates.md) section for information on how to implement postfix templates with more advanced features and editing possibilities.
+>
+{type="note"}
+
+## Postfix Template Description
+
+All postfix templates must provide descriptions and examples showing the code before and after a template is expanded.
+The files describing the template should be placed in the plugin's <path>resources</path> in the <path>postfixTemplates/_$TEMPLATE_NAME$_</path> where the _$TEMPLATE_NAME$_ directory must match the simple name of the template class,
+e.g., for a template implemented in `com.example.IntroduceVariablePostfixTemplate` class, the directory name should be named as <path>IntroduceVariablePostfixTemplate</path>.
+
+Providing the description explaining the template purpose and context details is achieved by creating the <path>description.html</path> file.
+
+Providing the code snippets showing the template in "before" and "after" expanding states is achieved via the <path>before._$EXTENSION$_.template</path> and <path>after._$EXTENSION$_.template</path> files accordingly.
+The _$EXTENSION$_ placeholder should be replaced with the extension of the template language, e.g., <path>before.kt.template</path> for a Kotlin template.
+
+The code snippets included in the example files can use the `<spot>` marker, which should surround the most important code parts, e.g., expression to expand and position of the caret after expanding.
+Marked parts will be highlighted in the <menupath>Settings/Preferences | Editor | General | Postfix Completion</menupath> settings page, making it easier for users to understand how a template is expanded, e.g.:
+- <path>before.java.template</path>:
+  ```java
+  <spot>cart.getProducts()</spot>.var
+  ```
+
+- <path>after.java.template</path>:
+  ```java
+  List<Product> products = cart.getProducts();<spot></spot>
+  ```
+
+Template example files can also use the `$key` placeholder which is replaced with the actual template key in the preview UI, e.g., consider a template with the `var` key:
+
+<compare title-before="Content" title-after="Rendition">
+  <code style="block" lang="Text">
+    cart.getProducts()$key
+  </code>
+  <code style="block" lang="Text">
+    cart.getProducts().var
+  </code>
+</compare>
+
+The gutter icons for a postfix template class allow navigating to the corresponding description and before/after files in plugin resources.
+
+**Example:**
+[`TryWithResourcesPostfixTemplate`](upsource:///java/java-impl/src/postfixTemplates/TryWithResourcesPostfixTemplate)
+directory containing description files for
+[`TryWithResourcesPostfixTemplate`](upsource:///java/java-impl/src/com/intellij/codeInsight/template/postfix/templates/TryWithResourcesPostfixTemplate.java)
+template.
```

Comment:
a short note about advanced page at end of this page?

Reply:
This is done in line 33.